### PR TITLE
[pulsar-java-client] Decouple partition switch frequency with maxPublishDelay in producer configuration

### PR DESCRIPTION
**Pull request by yjshen**
# Motivation

Currently, while using round-robin message router, messages would be route to different partitions at the `maxPublishDelay` boundaries. However, when GC occurs, the output partition would change frequently and results in small batches. 

This PR introduces a conf field setting the round-robin partition switch frequency: `batchingPartitionSwitchFrequencyByPublishDelay`. With this PR, the partition switch interval could be set as `maxPublishDelay` * `frequency`.

### Modifications

`batchingPartitionSwitchFrequencyByPublishDelay` in `ProducerConfiguratiionData.java`

### Verifying this change

This change is already covered by existing tests.

### Does this pull request potentially affect one of the following parts:
  - The public API: (yes)

### Documentation

  - Does this pull request introduce a new feature? (yes )
  - If yes, how is the feature documented? (docs / JavaDocs)

### CODE REVIEW

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerBuilder.java [hunk 1]
```diff
@@ -349,6 +349,20 @@
...
+     *
+     * The time period of partition switch is frequency * batchingMaxPublishDelay. During this period,
+     * all messages arrives will be route to the same partition.
+     *
+     * @param frequency the frequency of partition switch
+     * @return the producer builder instance
+     * @see #messageRoutingMode(MessageRoutingMode)
+     * @see #batchingMaxPublishDelay(long, TimeUnit)
+     */
+    ProducerBuilder<T> batchingPartitionSwitchFrequency(int frequency);
```
Comment (sijie) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerBuilder.java]: It is better to add both time interval and time unit in the method, by following the convention of other time related settings.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/conf/ProducerConfigurationData.java [hunk 2]
```diff
@@ -70,6 +70,7 @@
     private MessageRouter customMessageRouter = null;
 
     private long batchingMaxPublishDelayMicros = TimeUnit.MILLISECONDS.toMicros(1);
+    private int batchingPartitionSwitchFrequencyByPublishDelay = 10;
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/conf/ProducerConfigurationData.java]: Add "Micros" to the name so people know the time unit directly by reading the name.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerBuilder.java (hunk 1 above)
Comment (yjshen) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerBuilder.java]: The current frequency is used to compute switch delay in `frequency` * `maxBatchDelayMicros`. Do you mean I should make it a standalone variable? Then if users didn't set this variable, or the user is setting a variable less than `maxBatchDelay`, I need to use `maxBatchDelay` in round-robin partitioner?

Comment (sijie) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerBuilder.java]: Oh I misread the code. I see it now.  Then that's fine to me.

**Comment by yjshen** [comment]
> The change looks good me, is it possible implement in `RoundRobinPartitionMessageRouterImpl`? Looks the batching partition switch frequency only works in RoundRobin mode. Please considering whether this configuration should in ProducerBuilder.

I think we are exposing the capability of changing partition switch frequency to users? If so, the user conf in `RoundRobinPartitionMessageRouterImpl` should be fetched from `ProducerConfigurationData`?

**Comment by yjshen** [comment]
run cpp tests.

**Comment by codelipenghui** [comment]
@sijie I have discussed with yijie, i'm not sure use a custom router is reasonable here. we can support setting frequency in RoundRobinPartitionMessageRouterImpl and user use it like:

```
newProducer().router(new RoundRobinPartitionMessageRouterImpl(int frequency))
```

Which can reduces `ProducerConfigurationData` complexity since we all ready have  routeMode and customRouter there.

**Comment by sijie** [comment]
@codelipenghui that works for me.

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by yjshen** [comment]
run java8 tests.
run integration tests.

**Comment by yjshen** [comment]
run java8 tests.
run integration tests.

**Comment by yjshen** [comment]
run java8 tests.
run integration tests.

**Pull request merged**